engine/repositories/assistant_fs_repository.cc

```diff
@@ -0,0 +1,207 @@
+      }
+
+      std::shared_lock assistant_lock(GrabAssistantMutex(current_assistant_id));
+      auto assistant_res = LoadAssistant(current_assistant_id);
+      if (assistant_res.has_value()) {
+        all_assistants.push_back(std::move(assistant_res.value()));
+      }
+      assistant_lock.unlock();
+    }
+
+    // sorting
+    if (order == "desc") {
+      std::sort(all_assistants.begin(), all_assistants.end(),
+                [](const OpenAi::Assistant& assistant1,
+                   const OpenAi::Assistant& assistant2) {
+                  return assistant1.created_at > assistant2.created_at;
+                });
+    } else {
+      std::sort(all_assistants.begin(), all_assistants.end(),
+                [](const OpenAi::Assistant& assistant1,
+                   const OpenAi::Assistant& assistant2) {
+                  return assistant1.created_at < assistant2.created_at;
+                });
+    }
+
+    size_t assistant_count =
+        std::min(static_cast<size_t>(limit), all_assistants.size());
+    for (size_t i = 0; i < assistant_count; i++) {
+      assistants.push_back(std::move(all_assistants[i]));
+    }
+
+    return assistants;
+  } catch (const std::exception& e) {
+    return cpp::fail("Failed to list assistants: " + std::string(e.what()));
+  }
+}
+
+cpp::result<OpenAi::Assistant, std::string>
+AssistantFsRepository::RetrieveAssistant(const std::string assistant_id) const {
+  std::shared_lock lock(GrabAssistantMutex(assistant_id));
+  return LoadAssistant(assistant_id);
+}
+
+cpp::result<void, std::string> AssistantFsRepository::ModifyAssistant(
+    OpenAi::Assistant& assistant) {
+  std::unique_lock lock(GrabAssistantMutex(assistant.id));
+  auto path = GetAssistantPath(assistant.id);
+
+  if (!std::filesystem::exists(path)) {
+    return cpp::fail("Assistant doesn't exist: " + assistant.id);
+  }
+
+  return SaveAssistant(assistant);
+}
+
+cpp::result<void, std::string> AssistantFsRepository::DeleteAssistant(
+    const std::string& assitant_id) {
+  {
+    std::unique_lock assistant_lock(GrabAssistantMutex(assitant_id));
+    auto path = GetAssistantPath(assitant_id);
+    if (!std::filesystem::exists(path)) {
+      return cpp::fail("Assistant doesn't exist: " + assitant_id);
+    }
+    try {
+      std::filesystem::remove_all(path);
+    } catch (const std::exception& e) {
+      return cpp::fail("");
+    }
+  }
+
+  std::unique_lock map_lock(map_mutex_);
+  assistant_mutexes_.erase(assitant_id);
+  return {};
+}
+
+cpp::result<OpenAi::Assistant, std::string>
+AssistantFsRepository::CreateAssistant(OpenAi::Assistant& assistant) {
+  std::unique_lock lock(GrabAssistantMutex(assistant.id));
+  auto path = GetAssistantPath(assistant.id);
+
+  if (std::filesystem::exists(path)) {
+    return cpp::fail("Assistant already exists: " + assistant.id);
+  }
+
+  std::filesystem::create_directories(path);
+  auto assistant_file_path = path / kAssistantFileName;
+  std::ofstream assistant_file(assistant_file_path);
```

Comment:
Is it necessary since we create file if it does not exist in `SaveAssistant` function?

Reply:
It's needed because we need to create `assistant.json` inside the folder. ModifyAssistant also use `SaveAssistant` so can't move the create `assistant.json` there.